src/org/opensolaris/opengrok/web/api/v1/filter/AuthorizationFilter.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * See LICENSE.txt included in this distribution for the specific
+ * language governing permissions and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at LICENSE.txt.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved.
+ */
+package org.opensolaris.opengrok.web.api.v1.filter;
+
+import org.opensolaris.opengrok.authorization.AuthorizationFramework;
+import org.opensolaris.opengrok.configuration.Project;
+import org.opensolaris.opengrok.configuration.RuntimeEnvironment;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerRequestFilter;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.ext.Provider;
+import java.util.Map;
+
+@Provider
+@Authorized
+public class AuthorizationFilter implements ContainerRequestFilter {
+
+    public static final String PROJECTS_PARAM = "projects";
+
+    private final RuntimeEnvironment env = RuntimeEnvironment.getInstance();
+
+    @Context
+    private HttpServletRequest request;
+
+    @Override
+    public void filter(final ContainerRequestContext context) {
+        AuthorizationFramework auth = env.getAuthorizationFramework();
+        if (auth != null) {
+            for (String project : getProjects()) {
+                Project p = Project.getByName(project);
+                if (!auth.isAllowed(request, p)) {
+                    context.abortWith(Response.status(Response.Status.FORBIDDEN).build());
+                }
+            }
+        }
+    }
+
+    private String[] getProjects() {
+        String[] projects = request.getParameterValues(PROJECTS_PARAM);
+        if (projects != null) {
+            return projects;
+        }
+
+        // if no projects are specified then it means that all projects are wanted
+        Map<String, Project> projectsMap = env.getProjects();
+        if (projectsMap != null) {
+            return projectsMap.keySet().toArray(new String[0]);
```

Comment:
`new String[projectsMap.size()]`?

Reply:
I used to do it as well. Only recently IDEA started to suggest:

> There are two styles to convert a collection to an array: either using a pre-sized array (like c.toArray(new String[c.size()])) or using an empty array (like c.toArray(new String[0]).
In older Java versions using pre-sized array was recommended, as the reflection call which is necessary to create an array of proper size was quite slow. However since late updates of OpenJDK 6 this call was intrinsified, making the performance of the empty array version the same and sometimes even better, compared to the pre-sized version. Also passing pre-sized array is dangerous for a concurrent or synchronized collection as a data race is possible between the size and toArray call which may result in extra nulls at the end of the array, if the collection was concurrently shrunk during the operation.
This inspection allows to follow the uniform style: either using an empty array (which is recommended in modern Java) or using a pre-sized array (which might be faster in older Java versions or non-HotSpot based JVMs).

Reply:
ok